pkg/utils/test/matchers/managedresource.go

```diff
@@ -0,0 +1,101 @@
+// Copyright 2024 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package matchers
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/onsi/gomega/format"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+
+	resourcesv1alpha1 "github.com/gardener/gardener/pkg/apis/resources/v1alpha1"
+)
+
+type managedResourceDataMatcher struct {
+	cl          client.Client
+	expectedObj client.Object
+
+	expectedObjectSerialized string
+	actualObjectSerialized   string
+	secretNames              []string
+}
+
+func (m *managedResourceDataMatcher) FailureMessage(_ interface{}) (message string) {
+	if m.actualObjectSerialized != "" {
+		return format.Message(m.actualObjectSerialized, "to equal", m.expectedObjectSerialized)
+	}
+	return format.Message(m.expectedObjectSerialized, fmt.Sprintf("to be found in managed resource secrets %v", m.secretNames))
+}
+
+func (m *managedResourceDataMatcher) NegatedFailureMessage(_ interface{}) (message string) {
+	return format.Message(m.expectedObjectSerialized, fmt.Sprintf("not to be found in managed resource secrets %v", m.secretNames))
+}
+
+func (m *managedResourceDataMatcher) Match(actual interface{}) (bool, error) {
+	if actual == nil {
+		return false, nil
+	}
```

Comment:
Not sure if I understand the logic completely, but can a match `nil == nil` not happen? In other words, can the expected value never be `nil` so that the match cannot occur?

Reply:
As discussed, the approach is followed by error matcher https://github.com/gardener/gardener/blob/ce217c27e1fe3547baf585c43abc2b230fea1c12/pkg/utils/test/matchers/kubernetes_errors.go#L30.